dashboard/app/controllers/projects_controller.rb

```diff
@@ -571,6 +571,42 @@ def limited_gallery?
     !project_validator && limited_project_gallery
   end
 
+  # GET /projects/:channel_id/extra_links
+  # Get the extra links for the project for use by project validators.
+  # This is used by lab2 levels that cannot use the haml 'extra links' box since
+  # this box will not refresh when changing levels.
+  def extra_links
+    src_channel_id = params[:channel_id]
+    if src_channel_id == "undefined"
+      return render json: {message: 'No channel id provided.'}, status: :ok
+    end
+    project_info = {}
+    owner_info = {}
+    owner_info['storage_id'], project_info['id'] = storage_decrypt_channel_id(src_channel_id)
+    project_info['sources_link'] = "https://s3.console.aws.amazon.com/s3/buckets/#{CDO.sources_s3_bucket}/#{CDO.sources_s3_directory}/#{owner_info['storage_id']}/#{project_info['id']}/"
```

Comment:
Out of curiosity - is this how we construct S3 links for the legacy extra links box as well?

Reply:
yes - https://github.com/code-dot-org/code-dot-org/blob/a0f1ed29a6b217c930590a020159b4b77bfcace5/dashboard/app/views/levels/_admin.html.haml#L123